### PR TITLE
refactor: Lazy initialization of clients (admin, sr, ksql, connect)

### LINKED ISSUE #3663: pull queries open a kafka admin client per request
**Describe the bug**
spent a bunch of time profiling the ksql server further to understand (I still don’t know how to control min/max threads in jetty), but nonetheless I saw that admin client threads are being generated per request. (can repro this even with show tables, so its more to do with the /ksql endpoint than anything related to pull queries)

**To Reproduce**
Steps to reproduce the behavior, include:
1. master branch
2. N/A
3. Just do a curl of `SHOW TABLES` in a loop

**Expected behavior**
Kafka admin client should not be opened per request? 

**Actual behaviour**

ksql-with-pull-queries

```
2019-10-23 17:20:36,607] INFO Received: KsqlRequest{ksql='SHOW TABLES;', streamsProperties={}, commandSequenceNumber=Optional.empty} (io.confluent.ksql.rest.server.resources.KsqlResource:198)
java.lang.RuntimeException: Whos creating this?qtp843710487-42
    at org.apache.kafka.clients.admin.KafkaAdminClient.<init>(KafkaAdminClient.java:508)
    at org.apache.kafka.clients.admin.KafkaAdminClient.createInternal(KafkaAdminClient.java:456)
    at org.apache.kafka.clients.admin.Admin.create(Admin.java:71)
    at org.apache.kafka.streams.processor.internals.DefaultKafkaClientSupplier.getAdmin(DefaultKafkaClientSupplier.java:41)
    at io.confluent.ksql.services.ServiceContextFactory.create(ServiceContextFactory.java:53)
    at io.confluent.ksql.rest.server.services.RestServiceContextFactory.create(RestServiceContextFactory.java:71)
    at io.confluent.ksql.rest.server.services.RestServiceContextFactory.create(RestServiceContextFactory.java:57)
    at io.confluent.ksql.rest.server.context.KsqlRestServiceContextFactory.provide(KsqlRestServiceContextFactory.java:91)
    at io.confluent.ksql.rest.server.context.KsqlRestServiceContextFactory.provide(KsqlRestServiceContextFactory.java:39)
    at org.jvnet.hk2.internal.FactoryCreator.create(FactoryCreator.java:129)
```

**Additional context**

### PR DESCRIPTION
**Pull request by vpapavas**
### Description 
Fixes #3663 

When RBAC is not enabled, a new service context with kafka, sr, ksql clients is created PER request. With this PR, we will reuse the default SchemaContext created when the ksql REST application starts in the case of not enabled RBAC.

Open issues that need to resolved after the release: Should we make the `SchemaRegistryClient` lazy as well? Issue #3852 

### Testing done 
Added tests to `StreamedQueryResourceTest` and `MemoizedSupplierTest`

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java [hunk 1]
```diff
@@ -163,7 +162,7 @@ public static SourceName getCommandsStreamName() {
       final StreamedQueryResource streamedQueryResource,
       final KsqlResource ksqlResource,
       final VersionCheckerAgent versionCheckerAgent,
-      final BiFunction<KsqlConfig, KsqlSecurityExtension, Binder> serviceContextBinderFactory,
+      final KsqlRestServiceContextBinderFunction serviceContextBinderFactory,
```
Comment (vinothchandar) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java]: rename to `serviceContextBinderFunction` ?

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/KsqlResource.java [hunk 2]
```diff
@@ -208,13 +208,15 @@ public Response handleKsqlStatements(
...
 
       final List<ParsedStatement> statements = ksqlEngine.parse(request.getKsql());
-      validator.validate(
-          SandboxedServiceContext.create(serviceContext),
-          statements,
-          request.getStreamsProperties(),
-          request.getKsql()
-      );
 
+      if (!request.getKsql().toUpperCase().startsWith("SELECT")) {
```
Comment (vinothchandar) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/resources/KsqlResource.java]: ah this is still needed to avoid httpClientBuilder overhead? Should we consider a better way of special casing pull query? looks like we cannot avoid it. ?

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java (hunk 1 above)
Comment (big-andy-coates) [reply ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/KsqlRestApplication.java]: Or maybe better to rename `KsqlRestServiceContextBinderFunction` to `KsqlRestServiceContextBinderFactory` as it is a factor

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/context/KsqlRestServiceContextFactory.java [hunk 3]
```diff
@@ -37,13 +37,16 @@
...
+  private static ServiceContext serviceContext;
   private static KsqlConfig ksqlConfig;
   private static KsqlSecurityExtension securityExtension;
 
   public static void configure(
+      final ServiceContext serviceContext,
       final KsqlConfig ksqlConfig,
       final KsqlSecurityExtension securityExtension
   ) {
+    KsqlRestServiceContextFactory.serviceContext = serviceContext;
```
Comment (big-andy-coates) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/context/KsqlRestServiceContextFactory.java]: nit: null check

**Comment by vpapavas** [comment]
>This PR doesn't seem to do this, or am I missing something?

You are right, it is not. The description was outdated. What this PR is attempting to do is to not create a new `ServiceContext` per request but rather use the default one created at startup.

>I think a better approach to the one taken in this PR would be to make the instantiation of the admin client, and other clients for external services, lazy, i.e. only create them on first use. As pull queries against materialized state don't interact with Kafka, this will mean no clients are created.

You are right and this is exactly what I tried to do here.

**Comment by vpapavas** [comment]
I made the client creation lazy by using suppliers and memoizing them. The current implementation is not thread-safe however this should not be an issue as a request is a single thread. 

Disclaimer: I tried using guava `Suppliers.memoize` but the problem was that I need to be able to check if a client has been initialized and then only close it in `DefaultServiceContext.close`. Guava doesn't expose this check and always calls the supplier if it is not initialized.

I profiled and there are no `AdminClients` or `KsqlClient` being built.

File: ksql-engine/src/main/java/io/confluent/ksql/services/ClientContextFactory.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+package io.confluent.ksql.services;
+
+import io.confluent.kafka.schemaregistry.client.SchemaRegistryClient;
+import java.util.function.Supplier;
+import org.apache.kafka.clients.admin.Admin;
+
+public class ClientContextFactory {
+
+  private final MemoizedSupplier<Admin> memoizedAdminClientSupplier;
+  private MemoizedSupplier<ConnectClient> memoizedConnectClientSupplier;
```
Comment (vinothchandar) [inline ksql-engine/src/main/java/io/confluent/ksql/services/ClientContextFactory.java]: should these 3 be also final?

File: ksql-engine/src/main/java/io/confluent/ksql/services/ClientContextFactory.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+  public MemoizedSupplier<SchemaRegistryClient> getMemoizedSRClientSupplier() {
+    return memoizedSRClientSupplier;
+  }
+
+  public MemoizedSupplier<SimpleKsqlClient> getMemoizedKsqlClientSupplier() {
+    return memoizedKsqlClientSupplier;
+  }
+
+  public static class MemoizedSupplier<T> {
```
Comment (vinothchandar) [inline ksql-engine/src/main/java/io/confluent/ksql/services/ClientContextFactory.java]: I was expecting this to be a `Supplier` implementation as well. Should we do that since you have a `get()` anyway? This way, it could play well with java APIs that work with Supplier in the future.

I understand you have to pass it around as `MemoizedSupplier` still, for the init check,

File: ksql-engine/src/main/java/io/confluent/ksql/services/ServiceContextFactory.java [hunk 6]
```diff
@@ -48,19 +48,19 @@ public static ServiceContext create(
...
       final Supplier<SchemaRegistryClient> srClientFactory,
       final ConnectClient connectClient,
-      final SimpleKsqlClient ksqlClient
+      final Supplier<SimpleKsqlClient> ksqlClientSupplier
   ) {
-    final Admin adminClient = kafkaClientSupplier.getAdmin(
-        ksqlConfig.getKsqlAdminClientConfigProps()
-    );
+
+    ClientContextFactory clientsContext = new ClientContextFactory(
```
Comment (vinothchandar) [inline ksql-engine/src/main/java/io/confluent/ksql/services/ServiceContextFactory.java]: So the decision here is to still create the ClientContextFactory every request and hope its cheap? 

I am fine with this for now. But we will have revisit all this in the context of caching in general for RBAC..

Comment (vpapavas) [reply ksql-engine/src/main/java/io/confluent/ksql/services/ServiceContextFactory.java]: This should be cheap as it is just keeping the suppliers. No object is being created, until actually needed. And because of memoization, it is created only once per request. For Pull queries, none should be created.

File: ksql-engine/src/main/java/io/confluent/ksql/embedded/KsqlContext.java [hunk 7]
```diff
@@ -76,7 +76,7 @@ public static KsqlContext create(
   ) {
     Objects.requireNonNull(ksqlConfig, "ksqlConfig cannot be null.");
     final ServiceContext serviceContext =
-        ServiceContextFactory.create(ksqlConfig, DisabledKsqlClient.instance());
+        ServiceContextFactory.create(ksqlConfig, () -> DisabledKsqlClient.instance());
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/embedded/KsqlContext.java]:
```suggestion
        ServiceContextFactory.create(ksqlConfig, DisabledKsqlClient::instance);
```

File: ksql-engine/src/main/java/io/confluent/ksql/services/ClientContextFactory.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    public T get() {
+      if (!isInitialized) {
+        value = supplier.get();
+        isInitialized = true;
+      }
+      return value;
+    }
+
+    public boolean isInitialized() {
+      return isInitialized;
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/services/ClientContextFactory.java]: no need for `isInitialized` boolean.  Just do:

```suggestion
      return value != null;
```

File: ksql-engine/src/main/java/io/confluent/ksql/services/ClientContextFactory.java [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+  public MemoizedSupplier<SchemaRegistryClient> getMemoizedSRClientSupplier() {
+    return memoizedSRClientSupplier;
+  }
+
+  public MemoizedSupplier<SimpleKsqlClient> getMemoizedKsqlClientSupplier() {
+    return memoizedKsqlClientSupplier;
+  }
+
+  public static class MemoizedSupplier<T> {
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/services/ClientContextFactory.java]: Can we rename this class `NonThreadSafeMemoizedSupplier` to make it clear its not thread safe?

Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/services/ClientContextFactory.java]: This class would be better suited as a implementation detail.  The only reason you expose it is so that you can close the admin client.  However, a cleaner pattern would be to add a `close` method to `ClientContextFactory`.  This would offer better encapsulation.

However, before making this change read my other comments, as I actually think we don't need `ClientContextFactory` at all.

Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/services/ClientContextFactory.java]: Can you make `final` please?

File: ksql-engine/src/main/java/io/confluent/ksql/services/KafkaTopicClientImpl.java [hunk 10]
```diff
@@ -70,15 +71,15 @@
...
 
   /**
    * Construct a topic client from an existing admin client.
    *
-   * @param adminClient the admin client.
+   * @param adminClientSupplier the admin client supplier.
    */
-  public KafkaTopicClientImpl(final Admin adminClient) {
-    this.adminClient = Objects.requireNonNull(adminClient, "adminClient");
+  public KafkaTopicClientImpl(final MemoizedSupplier<Admin> adminClientSupplier) {
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/services/KafkaTopicClientImpl.java]: There's no reason for this class to _know_ about `MemoizedSupplier`.  Rather than have it accept a `MemoizedSupplier`, just instantiate this class lazily as well. When you need to instantiate it, first instantiate the `Admin` client.

File: ksql-engine/src/test/java/io/confluent/ksql/embedded/KsqlContextTestUtil.java [hunk 11]
```diff
@@ -48,10 +50,18 @@ public static KsqlContext create(
   ) {
     final KafkaClientSupplier clientSupplier = new DefaultKafkaClientSupplier();
 
-    final Admin adminClient = clientSupplier
-        .getAdmin(ksqlConfig.getKsqlAdminClientConfigProps());
+    ClientContextFactory clientContextFactory = new ClientContextFactory(
```
Comment (big-andy-coates) [inline ksql-engine/src/test/java/io/confluent/ksql/embedded/KsqlContextTestUtil.java]: You should be able to revert the changes in this file with my suggested approach

File: ksql-engine/src/test/java/io/confluent/ksql/integration/JsonFormatTest.java [hunk 12]
```diff
@@ -95,7 +95,7 @@ public void before() throws Exception {
     streamName = "STREAM_" + COUNTER.getAndIncrement();
 
     ksqlConfig = KsqlConfigTestUtil.create(CLUSTER);
-    serviceContext = ServiceContextFactory.create(ksqlConfig, DisabledKsqlClient.instance());
+    serviceContext = ServiceContextFactory.create(ksqlConfig, () -> DisabledKsqlClient.instance());
```
Comment (big-andy-coates) [inline ksql-engine/src/test/java/io/confluent/ksql/integration/JsonFormatTest.java]:
```suggestion
    serviceContext = ServiceContextFactory.create(ksqlConfig, DisabledKsqlClient::instance);
```

File: ksql-engine/src/test/java/io/confluent/ksql/integration/SecureIntegrationTest.java [hunk 13]
```diff
@@ -114,18 +116,17 @@
   private final TopicProducer topicProducer = new TopicProducer(SECURE_CLUSTER);
   private KafkaTopicClient topicClient;
   private String outputTopic;
-  private AdminClient adminClient;
+  private MemoizedSupplier<Admin> adminClient;
```
Comment (big-andy-coates) [inline ksql-engine/src/test/java/io/confluent/ksql/integration/SecureIntegrationTest.java]: You should be able to revert these changes with my suggested approach.

File: ksql-engine/src/test/java/io/confluent/ksql/integration/SecureIntegrationTest.java [hunk 14]
```diff
@@ -266,7 +267,7 @@ private static void givenAllowAcl(final Credentials credentials,
 
   private void givenTestSetupWithConfig(final Map<String, Object> ksqlConfigs) {
     ksqlConfig = new KsqlConfig(ksqlConfigs);
-    serviceContext = ServiceContextFactory.create(ksqlConfig, DisabledKsqlClient.instance());
+    serviceContext = ServiceContextFactory.create(ksqlConfig, () -> DisabledKsqlClient.instance());
```
Comment (big-andy-coates) [inline ksql-engine/src/test/java/io/confluent/ksql/integration/SecureIntegrationTest.java]:
```suggestion
    serviceContext = ServiceContextFactory.create(ksqlConfig, DisabledKsqlClient::instance);
```

File: ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/TestKsqlRestApp.java [hunk 15]
```diff
@@ -414,7 +414,7 @@ private static ServiceContext defaultServiceContext(
     final KsqlConfig config =
         new KsqlConfig(buildConfig(bootstrapServers, baseConfig).getKsqlConfigProperties());
 
-    return ServiceContextFactory.create(config, DisabledKsqlClient.instance());
+    return ServiceContextFactory.create(config, () -> DisabledKsqlClient.instance());
```
Comment (big-andy-coates) [inline ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/TestKsqlRestApp.java]:
```suggestion
    return ServiceContextFactory.create(config, DisabledKsqlClient::instance);
```

File: ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/StandaloneExecutorFactory.java [hunk 16]
```diff
@@ -57,7 +57,7 @@ public static StandaloneExecutor create(
         properties,
         queriesFile,
         installDir,
-        config -> ServiceContextFactory.create(config, DisabledKsqlClient.instance()),
+        config -> ServiceContextFactory.create(config, () -> DisabledKsqlClient.instance()),
```
Comment (big-andy-coates) [inline ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/StandaloneExecutorFactory.java]:
```suggestion
        config -> ServiceContextFactory.create(config, DisabledKsqlClient::instance),
```

**Comment by vpapavas** [comment]
After applying all the changes and testing the code, a new AdminClient is created per request when it shouldn't since it is not used by anything. With my initial changes, there was none created. 

The problem is the `DefaultServiceContext.getTopicClient`.  This is because with the change you requested, first it will call `getAdminClient` which will create the `AdminClient` and then do`getTopicClient` which calls `KafkaTopicClientImpl`. The `AdminClient` is created only to be passed as argument, it is not used. 

That's why I had changed the signature of `KafkaTopicClientImpl` to take a supplier as argument.

File: ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java [hunk 17]
```diff
@@ -70,26 +120,56 @@ public KafkaClientSupplier getKafkaClientSupplier() {
...
 
   @Override
   public void close() {
-    adminClient.close();
+    if (((NotThreadSafeMemoizedSupplier)adminClientSupplier).isInitialized()) {
+      adminClientSupplier.get().close();
+    }
+  }
+
+  public static final class NotThreadSafeMemoizedSupplier<T> implements Supplier<T> {
```
Comment (agavra) [inline ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java]: is it OK for this to be thread unsafe? If I understand correctly, and @spena can probably confirm, but I think we use `DefaultServiceContext` from multiple threads (we create one per request for the "user", but we also store one for the "server"). Either way, I think thread safety (using guava's memoized supplier) won't cause a noticeable performance overhead - a volatile field lookup in the average case is basically free - so we may as well do that unless we have justification otherwise

File: ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java [hunk 18]
```diff
@@ -29,38 +30,87 @@
...
     this.kafkaClientSupplier = requireNonNull(kafkaClientSupplier, "kafkaClientSupplier");
-    this.adminClient = requireNonNull(adminClient, "adminClient");
-    this.topicClient = requireNonNull(topicClient, "topicClient");
-    this.srClientFactory = requireNonNull(srClientFactory, "srClientFactory");
-    this.srClient = requireNonNull(srClientFactory.get(), "srClient");
-    this.connectClient = requireNonNull(connectClient, "connectClient");
-    this.ksqlClient = requireNonNull(ksqlClient, "ksqlClient");
+
+    requireNonNull(topicClient, "topicClient");
+    this.topicClientSupplier = new NotThreadSafeMemoizedSupplier<KafkaTopicClient>(
```
Comment (agavra) [inline ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java]: nit: this constructor should just call the other one

File: ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java [hunk 19]
```diff
@@ -29,38 +30,87 @@
 public class DefaultServiceContext implements ServiceContext {
 
   private final KafkaClientSupplier kafkaClientSupplier;
-  private final Admin adminClient;
-  private final KafkaTopicClient topicClient;
-  private final Supplier<SchemaRegistryClient> srClientFactory;
+  private final Supplier<Admin> adminClientSupplier;
+  private final Supplier<KafkaTopicClient>  topicClientSupplier;
+  private final Supplier<SchemaRegistryClient> srClientSupplier;
   private final SchemaRegistryClient srClient;
```
Comment (vinothchandar) [inline ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java]: why do we need this field also, if there is a supplier already?

File: ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java [hunk 20]
```diff
@@ -29,38 +30,87 @@
...
+
+    requireNonNull(connectClientSupplier, "connectClientSupplier");
+    this.connectClientSupplier = new NotThreadSafeMemoizedSupplier<ConnectClient>(
+        connectClientSupplier);
+
+    requireNonNull(ksqlClientSupplier, "ksqlClientSupplier");
+    this.ksqlClientSupplier = new NotThreadSafeMemoizedSupplier<SimpleKsqlClient>(
+        ksqlClientSupplier);
+
+    this.srClient = requireNonNull(srClientSupplier.get(), "srClient");
```
Comment (vinothchandar) [inline ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java]: is nt this non lazy? Is it because its a cheap object?

File: ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java (hunk 18 above)
Comment (vinothchandar) [reply ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java]: +1

File: ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java (hunk 17 above)
Comment (vinothchandar) [reply ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java]: +1 also.

Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java]: Good point. I think the reason @vpapavas created her own class was because we need to close the AdminClient.  If using Guava's version we'd have to `.get()` the instance to close it, which would create the instance if it hadn't already been created!  Basically, the issue is that there is no way to _know_ the instance had been created.  

I think a better solution would be to wrap the thread-safe guava memorizing supplier with something that tracks if its been initialized.  I'll knock up an example and add to my review.

File: ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java (hunk 18 above)
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java]: +1.

Though this is a little tricky. I'll make the change and post in my review.

File: ksql-engine/src/main/java/io/confluent/ksql/services/ServiceContextFactory.java [hunk 21]
```diff
@@ -48,19 +48,16 @@ public static ServiceContext create(
...
     return new DefaultServiceContext(
         kafkaClientSupplier,
-        adminClient,
-        new KafkaTopicClientImpl(adminClient),
+        () -> kafkaClientSupplier
+            .getAdmin(ksqlConfig.getKsqlAdminClientConfigProps()),
         srClientFactory,
-        connectClient,
-        ksqlClient
+        () -> connectClient,
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/services/ServiceContextFactory.java]: We're still creating a connect client per request then?

We should either make this lazy, or not have `DefaultServiceContext` accept a supplier for this.  It's misleading to have one-per-call _and_ the supplier.

It would be trivial to change this method to take a `Supplier<ConnectClient>` and change the two places its called to pass a supplier.

File: ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java (hunk 17 above)
Comment (vpapavas) [reply ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java]: Yes, Andy's comment is correct. I talked with @spena about the thread safety issue and he said that there is only one thread per request. That's why I created the supplier as not thread safe.

File: ksql-engine/src/main/java/io/confluent/ksql/services/KafkaTopicClientImpl.java [hunk 22]
```diff
@@ -70,14 +71,14 @@
   private static final String DEFAULT_REPLICATION_PROP = "default.replication.factor";
   private static final String DELETE_TOPIC_ENABLE = "delete.topic.enable";
 
-  private final Admin adminClient;
+  private final Supplier<Admin> adminClient;
```
[the remark this answers is not in the document]
Comment (vpapavas) [reply ksql-engine/src/main/java/io/confluent/ksql/services/KafkaTopicClientImpl.java]: Yes, the issue is in the `SandboxedServiceContext` that creates a `KafkaTopicClient` but doesn't use it. This is true for `/ksql` resource and pull queries.

File: ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java (hunk 17 above)
Comment (agavra) [reply ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java]: My concern was less that there's one thread per request, but rather that the ServiceContextFactory is shared across requests. @sergio do we create a new factory for every request (i.e. is `@Inject` called every request, or just `provide()`)? This is where I don't really have a working knowledge of the dependency injection framework that we use... either way, if there's no lock contention then synchronization is almost free and if there is thread contention then we need it 😂

File: ksql-engine/src/main/java/io/confluent/ksql/services/ServiceContextFactory.java (hunk 21 above)
Comment (agavra) [reply ksql-engine/src/main/java/io/confluent/ksql/services/ServiceContextFactory.java]: Creating a connect client is basically free - I wouldn't worry about optimizing this so I would just pass it in without the supplier. 
```
  public DefaultConnectClient(
      final String connectUri,
      final Optional<String> authHeader
  ) {
    Objects.requireNonNull(connectUri, "connectUri");
    this.authHeader = Objects.requireNonNull(authHeader, "authHeader");

    try {
      this.connectUri = new URI(connectUri);
    } catch (URISyntaxException e) {
      throw new KsqlException(
          "Could not initialize connect client due to invalid URI: " + connectUri, e);
    }
  }
```

File: ksql-engine/src/main/java/io/confluent/ksql/services/KafkaTopicClientImpl.java [hunk 23]
```diff
@@ -70,14 +71,14 @@
...
 
   /**
    * Construct a topic client from an existing admin client.
    *
    * @param adminClient the admin client.
    */
-  public KafkaTopicClientImpl(final Admin adminClient) {
+  public KafkaTopicClientImpl(final Supplier<Admin> adminClient) {
```
[the remark this answers is not in the document]
Comment (agavra) [reply ksql-engine/src/main/java/io/confluent/ksql/services/KafkaTopicClientImpl.java]: in what sense is this shared? please call that out in the javadoc as well

File: ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java (hunk 19 above)
Comment (vpapavas) [reply ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java]: I created an issue #3852 and added the link to the description of this PR

File: ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java [hunk 24]
```diff
@@ -70,26 +120,56 @@ public KafkaClientSupplier getKafkaClientSupplier() {
...
   @Override
   public SimpleKsqlClient getKsqlClient() {
-    return ksqlClient;
+    return ksqlClientSupplier.get();
   }
 
   @Override
   public void close() {
-    adminClient.close();
+    if (((NotThreadSafeMemoizedSupplier)adminClientSupplier).isInitialized()) {
```
Comment (spena) [inline ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java]: What about using a boolean value `adminInitialized` that is set to true when `getAdminClient` is called? I think that would help going back to the Guava memoizer, right? Thus making this code simpler with the removal of the NotThreadSafeMemoizedSupplier.

File: ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java [hunk 25]
```diff
@@ -29,38 +32,85 @@
 public class DefaultServiceContext implements ServiceContext {
 
   private final KafkaClientSupplier kafkaClientSupplier;
-  private final Admin adminClient;
-  private final KafkaTopicClient topicClient;
+  private final MemoizedSupplier<Admin> adminClientSupplier;
+  private final MemoizedSupplier<KafkaTopicClient>  topicClientSupplier;
   private final Supplier<SchemaRegistryClient> srClientFactory;
```
Comment (vinothchandar) [inline ksql-engine/src/main/java/io/confluent/ksql/services/DefaultServiceContext.java]: could we rename this to `srClientSupplier` as well?

File: ksql-engine/src/main/java/io/confluent/ksql/services/KafkaTopicClientImpl.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/services/KafkaTopicClientImpl.java]: OK, might be worth switching the sandbox one over to be lazy too in a follow up PR.
